jcheck/src/main/java/org/openjdk/skara/jcheck/CopyrightCheck.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.skara.jcheck;
+
+import org.openjdk.skara.census.Census;
+import org.openjdk.skara.vcs.Commit;
+import org.openjdk.skara.vcs.ReadOnlyRepository;
+import org.openjdk.skara.vcs.openjdk.CommitMessage;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.time.Year;
+import java.util.List;
+import java.util.regex.Pattern;
+
+public class CopyrightCheck extends CommitCheck {
+
+    private final ReadOnlyRepository repo;
+
+    private final static Pattern COPYRIGHT_PATTERN = Pattern.compile(".*Copyright \\(c\\) (\\d{4})(?:, (\\d{4}))?, Oracle and/or its affiliates\\. All rights reserved\\.");
+
+    CopyrightCheck(ReadOnlyRepository repo) {
+        this.repo = repo;
+    }
+
+    @Override
+    Iterator<Issue> check(Commit commit, CommitMessage message, JCheckConfiguration conf, Census census) {
+        var metadata = CommitIssue.metadata(commit, message, conf, this);
+        var pattern = Pattern.compile(conf.checks().copyright().files());
+
+        var filesWithCopyrightFormatIssue = new ArrayList<String>();
+        var filesWithCopyrightYearIssue = new ArrayList<String>();
+        var filesWithCopyrightMissingIssue = new ArrayList<String>();
+
+        for (var diff : commit.parentDiffs()) {
+            for (var patch : diff.patches()) {
+                if (patch.target().path().isEmpty() || patch.isBinary()) {
+                    continue;
+                }
+                var path = patch.target().path().get();
+                if (pattern.matcher(path.toString()).matches()) {
+                    try {
+                        var lines = repo.lines(path, commit.hash()).orElse(List.of());
+                        var copyrightFound = false;
+                        for (String line : lines) {
+                            if (line.contains("Copyright (c)") && line.contains("Oracle")) {
```

Comment:
I think it would be nice if we could work with pairs of regular expressions. One for identifying a line that should match and one that verifies the format. This way we could make it possible to add more variants through configuration.

We could also consider not having any default and put the Oracle config in .jcheck/conf.

Reply:
I'm not sure its better to have Oracle in the configuration file. It kind of just highlights it as a special case. If/when some other company requests special copyright checks, then we can generalize and move it to the configuration file. Also, it's never the case that this is just a per-repo setting: for Oracle-copyrighted code, this rule always applies.

Reply:
I don't want defaults because once we set them, we have to live with them. Changing the default configuration is risky as it applies globally to all commits that have already been checked as well. I think it's generally a better design having the configuration in the affected repo for increased transparency on what is actually being checked.